docs/user-guide/02-convenience-features.md

```diff
@@ -0,0 +1,227 @@
+```
+struct Foo
+{
+    static int staticMethod(int a, int b)
+    {
+        return a + b;
+    }
+}
+```
+
+Static member functions are accessed the same way as other static members, via either the type name or an instance of the type:
+
+```hlsl
+int rs = Foo.staticMethod(a, b);
+```
+
+or
+
+```hlsl
+Foo foo;
+...
+int rs = foo.staticMethod(a,b);
+```
+
+### Mutability of member function
+
+For GPU performance considerations, the `this` argument in a member function is immutable by default. If you modify the content in `this` argument, the modification will be discarded after the call and does not affect the input object. If you intend to define a member function that mutates the object, use `[mutating]` attribute on the member function as shown in the following example.
+
+```hlsl
+struct Foo
+{
+    int count;
+    
+    [mutating]
+    void setCount(int x) { count = x; }
+
+    void setCount2(int x) { count = x; }
+}
+
+void test()
+{
+    Foo f;
+    f.setCount(1); // f.count is 1 after the call.
+    f.setCount2(2); // f.count is still 1 after the call.
+}
+```
+
+## Properties
+
+Properties provide a convienient way to access values exposed by a type, where the logic behind accessing the value is defined in `getter` and `setter` function pairs. Slang's `property` feature is similar to C# and Swift. 
+```C#
+struct MyType
+{
+    uint flag;
+
+    property uint highBits
+    {
+        get { return flag >> 16; }
+        set { flag = (flag & 0xFF) + (newValue << 16); }
+    }
+};
+```
+
+Or equivalently in a "modern" syntax:
+
+```C#
+struct MyType
+{
+    uint flag;
+
+    property highBits : uint
+    {
+        get { return flag >> 16; }
+        set { flag = (flag & 0xFF) + (newValue << 16); }
+    }
+};
+```
+
+You may also use an explicit parameter for the setter method:
+```C#
+property uint highBits
+{
+    set(uint x) { flag = (flag & 0xFF) + (x << 16);  }
+}
+```
+
+> #### Note ####
+> Slang currently does not support automatically synthesized `getter` and `setter` methods. For example,
+> the following code is not supported:
+> ```
+> property uint highBits {get;set;} // Not supported yet.
+> ```
+
+## Initializers
+> #### Note ####
+> The syntax for defining initializers are subject to future change.
+
+
+Slang supports defining initializers in `struct` types. You can write:
+```C#
+struct MyType
+{
+    int myVal;
+    __init(int a, int b)
+    {
+        myVal = a + b;
+    }
+}
+```
+
+You can use an initializer to construct a new instance by using the type name in a function call expression:
+```C#
+MyType instance = MyType(1,2);  // instance.myVal is 3.
+```
+
+You may also use C++ style initializer list to invoke an initializer:
+```C#
+MyType instance = {1, 2};
+```
+
+If an initializer does not define any parameters, it will be recognized as *default* initializer that will be automatically called at the definition of a variable:
+
+```C#
+struct MyType
+{
+    int myVal;
+    __init(i)
```

Comment:
First note: the `i` here seems to be an accident.

Reply:
Second note: I want us to be *very* careful about whether to advertise this feature. Right now, I think of our support for default initializers as a bit of hack.

Almost every single non-C++ language (including all older versions of HLSL) has made a consistent choice between one of two options:

1. A variable without an initial-value expression is uninitialized, *always*. Access to a potentially-uninitialized variable may be a static error. or not. Access to an uninitialized variable at runtime may be a dynamic error, or it might yield an undefined value, or it might result in the dreaded Undefined Behavior.
2. A variable without an initial-value expression is default-initialized, *always*. Numeric values are initialized to zero, and references are initialized to null.

To my knowledge, C++ is the only mainstream language that doesn't pick one of these two internally-consistent options, and I'm reluctant to drag Slang down that road just because dxc has used some C++ features in the way it exposes new pipeline features.

Current HLSL does not support constructors/initializers in user code, and there is (to my knowledge) exactly *one* type in the current HLSL standard library that makes use of default construction: `RayQuery<M>`. (There is also the `HitGroup` type that uses C++ variable construction syntax as part of its idiom, but it doesn't rely on *default* construction)

I am very hesitant to introduce this feature as something our users are expected to know about. I would rather flag that `RayQuery<M>` initializer as an extreme special case, and perhaps even warn users when they declare a `RayQuery<M>` without initializing it (and advise them to do `var query = RayQuery<M>();` or similar.

Let's plan to talk about this and come to a conclusion.